src/main/java/stats/KMeans.java

```diff
@@ -0,0 +1,213 @@
+     * @return K clusters along with their features.
+     */
+    public static Map<Centroid, List<Record>> fit(List<Record> records, int k, Distance distance, int maxIterations) {
+
+        applyPreconditions(records, k, distance, maxIterations);
+
+        List<Centroid> centroids = randomCentroids(records, k);
+        Map<Centroid, List<Record>> clusters = new HashMap<>();
+        Map<Centroid, List<Record>> lastState = new HashMap<>();
+
+        // Iterate for a pre defined number of times:
+        for (int i = 0; i < maxIterations; i++) {
+            boolean isLastIteration = i == maxIterations - 1;
+
+            // In each iteration find the nearest centroid for each record:
+            for (Record record : records) {
+                Centroid centroid = nearestCentroid(record, centroids, distance);
+                assignToCluster(clusters, record, centroid);
+            }
+            // If assignments remain the same, then we have nothing else to do:
+            boolean shouldTerminate = isLastIteration || clusters.equals(lastState);
+            lastState = clusters;
+            if (shouldTerminate) {
+                break;
+            }
+            // Relocate centroid after each iteration:
+            centroids = relocateCentroids(clusters);
+            clusters = new HashMap<>();
+        }
+        return lastState;
+
+    }
+
+    /**
+     * Generates k randomly placed centroids. First generate the possible value range
+     * for each attribute. Then generate random coordinates in the [min, max] range
+     * foe each attribute.
+     *
+     * @param records The data set.
+     * @param k       Number of clusters.
+     * @return Collection of randomly generated centroids.
+     */
+    public static List<Centroid> randomCentroids(List<Record> records, int k) {
+
+        List<Centroid> centroids = new ArrayList<>();
+        Map<String, Double> maxValues = new HashMap<>();
+        Map<String, Double> minValues = new HashMap<>();
+
+        for (Record record : records) {
+            record.getFeatures().forEach((key, value) -> {
+                // Compares the value with the current max and choose the bigger value:
+                maxValues.compute(key, (k1, max) -> max == null || value > max ? value : max);
+
+                // Compare the value with the current min and choose the smaller value:
+                minValues.compute(key, (k2, min) -> min == null || value < min ? value : min);
+            });
+        }
+
+        Set<String> attributes = records.stream().flatMap(e -> e.getFeatures().keySet().stream()).collect(toSet());
+        for (int i = 0; i < k; i++) {
+            Map<String, Double> coordinates = new HashMap<>();
+            for (String attribute : attributes) {
+                double max = maxValues.get(attribute);
+                double min = minValues.get(attribute);
+                coordinates.put(attribute, random.nextDouble() * (max - min) + min);
```

Comment:
Having the centroids be one of the data points sounds very reasonable...does this implementation guarantees this? Don't think so is simply shifts the generated point in the [min, max) range but this does not mean that there exists actually such a data point in the set

Reply:
Perhaps labelling the data points is a good idea and choose randomly from these labels??